File: events/year_2023/day_04/day_04.py
"""
Advent of Code - Year 2023 - Day 4
https://adventofcode.com/2023/day/4
"""


def preprocessing(puzzle_input: str) -> list[int]:
    """
    Parses the puzzle input and returns a list of scores representing the count of matching numbers
    for each game.
    """
    scores = []
    for game in puzzle_input.splitlines():
        draw, card = game.split(' | ')
        draw = [int(n) for n in draw.split()[2:]]
        card = [int(n) for n in card.split()]
        score = len([n for n in card if n in draw])
        scores.append(score)
    return scores


def solver(scores: list[int]) -> tuple[int, int]:
    """
    Calculates total points and the sum of scratch cards based on a list of scores.
    """
    points = 0
    scratch = [1 for _ in range(len(scores))]
    for game, score in enumerate(scores):
        points += (1 << score) >> 1
        for gg in range(game + 1, game + score + 1):
            scratch[gg] += scratch[game]
    return points, sum(scratch)

File: events/year_2023/day_04/day_04.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<usize> {
    input
        .lines()
        .map(|game| {
            game.split(&[':', '|'])
                .skip(1)
                .map(|numbers| {
                    numbers
                        .split_whitespace()
                        .map(|number| number.parse::<u8>().unwrap())
                        .collect::<HashSet<u8>>()
                })
                .collect_tuple()
                .unwrap()
        })
        .map(|(draw, card)| draw.intersection(&card).count())
        .collect_vec()
}

fn part_1(input: Vec<usize>) -> usize {
    input.into_iter().map(|n| (1 << n) >> 1).sum()
}

fn part_2(input: Vec<usize>) -> usize {
    input
        .iter()
        .enumerate()
        .fold(vec![1; input.len()], |mut scratch, (game, score)| {
            (game..game + score).for_each(|gg| scratch[gg + 1] += scratch[game]);
            scratch
        })
        .iter()
        .sum()
}
